### PR TITLE
Testing bounty model Phi-1.5 with mirrorred branch of velonica0:phi-1.5-new2

### PR DESCRIPTION
**Pull request by gwangTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/19415

Original author: @velonica0

Initial PR with forked branch: https://github.com/tenstorrent/tt-metal/pull/29609

### Problem description
Mirror branch to add community contributed model -- Phi-1.5 -- to TTT

Target arch: N150/N300

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] Check accuracy results against bounty requirements: https://github.com/tenstorrent/tt-metal/issues/19415, with both performance and accuracy optimizations:
- [ ] Check TTFT and t/s/u results against bounty requirements: https://github.com/tenstorrent/tt-metal/issues/19415, with both performance and accuracy optimizations:
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes
- [ ] T3K demo tests
- [ ] Local test of unit tests on IRD with N300
  - [ ] models/tt_transformers/tests/test_embedding.py
  - [ ] models/tt_transformers/tests/test_rms_norm.py
  - [ ] models/tt_transformers/tests/test_mlp.py
  - [ ] models/tt_transformers/tests/test_attention.py
  - [ ] models/tt_transformers/tests/test_attention_prefill.py
  - [ ] models/tt_transformers/tests/test_decoder.py
  - [ ] models/tt_transformers/tests/test_decoder_prefill.py

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mirror/velonica0/phi-1.5-new2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mirror/velonica0/phi-1.5-new2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:mirror/velonica0/phi-1.5-new2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mirror/velonica0/phi-1.5-new2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mirror/velonica0/phi-1.5-new2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mirror/velonica0/phi-1.5-new2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mirror/velonica0/phi-1.5-new2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mirror/velonica0/phi-1.5-new2)

### CODE REVIEW

File: models/tt_transformers/tt/decoder.py [hunk 1]
```diff
@@ -45,6 +46,8 @@ def __init__(
         self.current = 0
         self.model_config = args.get_model_config()
         self.is_mixture_of_experts = False
+        self.parallel_model = ["phi-1", "phi-1.5"]
```
Comment (gwangTT) [inline models/tt_transformers/tt/decoder.py]: This should be refactored into model_config.py so in this file, we could just use `self.is_parallel_model`

File: models/tt_transformers/tt/decoder.py [hunk 2]
```diff
@@ -110,32 +114,35 @@ def __init__(
...
-                weight_cache_path=None if args.dummy_weights else weight_cache_path,
-                weight_dtype=ttnn.bfloat16,
-                weight_key="ffn_norm",
-                is_distributed=self.args.is_distributed_norm,
-                add_unit_offset=self.args.rms_norm_add_unit_offset,
-                sharded_program_config=self.model_config["SHARDED_NORM_MLP_PRGM_CFG"],
-                sharded_output_config=self.model_config["SHARDED_MLP_INPUT_MEMCFG"],
-                ccl_topology=self.args.ccl_topology(),
+
+        if f"layers.{self.layer_num}.ffn_norm.weight" in state_dict:
```
Comment (gwangTT) [inline models/tt_transformers/tt/decoder.py]: It would be cleaner to refactor this conditional into model_config.py so there we could use `if args.has_ffn_norm:`

File: models/tt_transformers/tt/decoder.py [hunk 3]
```diff
@@ -110,32 +114,35 @@ def __init__(
...
+                args,
                 tt_ccl=self.tt_ccl,
-            ),
-            args,
-            tt_ccl=self.tt_ccl,
-            TG=args.is_galaxy,
-        )
+                TG=args.is_galaxy,
+            )
         if f"layers.{layer_num}.pre_feedforward_layernorm.weight" in state_dict:
```
Comment (gwangTT) [inline models/tt_transformers/tt/decoder.py:153]: similar comment as above about refactoring.

File: models/tt_transformers/tt/decoder.py [hunk 4]
```diff
@@ -231,77 +239,128 @@ def forward(
...
-                    self.mesh_device,
-                    tt_ccl=self.tt_ccl,
-                    cluster_axis=0,
-                    dim=3,
-                    num_reduce_scatter_links=self.args.num_reduce_scatter_links,
-                    num_all_gather_links=self.args.num_all_gather_links,
-                    topology=ttnn.Topology.Ring,
-                    memory_config=ttnn.DRAM_MEMORY_CONFIG,
-                    dtype=self.args.ccl_dtype,
+        if self.args.base_model_name not in self.parallel_model:
```
Comment (gwangTT) [inline models/tt_transformers/tt/decoder.py]: related to an earlier comment on `self.parallel_model`: this here could be cleaner after the refactoring, e.g., `self.is_parallel_model`

File: models/tt_transformers/tt/decoder.py [hunk 5]
```diff
@@ -217,6 +224,7 @@ def forward(
 
         # Norms take fractured inputs and output replicated across devices
         attn_in = self.attention_norm(x, mode)
+        x_new = x
```
Comment (gwangTT) [inline models/tt_transformers/tt/decoder.py]: This could be moved to the else branch below. If the model is not a parallel model, this line of code is not need.

File: models/tt_transformers/tt/decoder.py (hunk 2 above)
Comment (gwangTT) [reply models/tt_transformers/tt/decoder.py]: [34b3b51](https://github.com/tenstorrent/tt-metal/pull/31359/commits/34b3b5177329204a0b61092790a4e0befeebadaf)

File: models/tt_transformers/tt/decoder.py (hunk 1 above)
Comment (gwangTT) [reply models/tt_transformers/tt/decoder.py]: [34b3b51](https://github.com/tenstorrent/tt-metal/pull/31359/commits/34b3b5177329204a0b61092790a4e0befeebadaf)

File: models/tt_transformers/tt/decoder.py (hunk 4 above)
Comment (gwangTT) [reply models/tt_transformers/tt/decoder.py]: [34b3b51](https://github.com/tenstorrent/tt-metal/pull/31359/commits/34b3b5177329204a0b61092790a4e0befeebadaf)

File: models/tt_transformers/tt/decoder.py (hunk 5 above)
Comment (gwangTT) [reply models/tt_transformers/tt/decoder.py]: [34b3b51](https://github.com/tenstorrent/tt-metal/pull/31359/commits/34b3b5177329204a0b61092790a4e0befeebadaf)
